### PR TITLE
Add Pay for Order support in ECE

### LINKED ISSUE #8869: Integrate ECE Payment Processing for the Pay for Order Endpoint
Implement the Stripe Express Checkout Element (ECE) to process payments on the pay for order endpoint (pending payment orders). 

### Acceptance criteria:

- The ECE implementation must successfully process payments for orders in the pending payment status via the pay for order page.
- Ensure there are no regressions in Payment Request Buttons (PRB) when ECE is disabled.
- There are no regressions when no express checkout payment methods are active.

### PR DESCRIPTION
**Pull request by cesarcosta99**
Fixes #8869.

#### Changes proposed in this Pull Request

This PR is getting the Express Checkout Elements (ECE) rendered in the Pay for Order flow and introducing the payment processing implementation for the same flow.

#### Testing instructions

> [!IMPORTANT]
> This has been tested using [Google Pay](https://pay.google.com) since it is currently difficult to test Apple Pay due to Apple Pay's requirements for test cards.

> [!IMPORTANT]
> You must test these changes on a public facing URL using HTTPS. Use something like Jurassic Tube or ngrok to create a tunnel to your local environment.

**Setup**

Ensure ECE is disabled by setting the `_wcpay_feature_stripe_ece` option to `0`.

**Regression test: Payment request buttons work as expected when ECE is disabled**

1. As a merchant, navigate to **WooCommerce > Orders > Add order**.
2. Click on **Add item(s)**, then on **Add product(s)**.
3. Type in some product name, such as "Beanie" and select it.
4. Click on the **Add** button, and finally, on the **Create** button.
5. Copy the **Customer payment page** link.
6. As a customer, navigate to the link you copied.
7. Pay using the Payment Request Button (PRB) and make sure the payment is processed successfully.
8. Test paying with the PRB in the product page, in the cart and in the checkout page as well.

**Regression test: Checkout with ECE still works in the standard flows**

1. Enable ECE by setting the `_wcpay_feature_stripe_ece` option to `1`.
2. As a customer, test paying with the ECE in the product page, in the cart and in the checkout page.

**Regression test: Regular checkout works when no express checkout payment method is active**

1. As a merchant, navigate to **Payments > Settings**.
2. Disable all express checkouts and save changes.
3. As a customer, add a product to the cart, go to checkout and complete the order. Make sure the payment is processed successfully.

**Test: Checkout with ECE in Pay for Order works with WooPay disabled**

1. As a merchant, navigate to **Payments > Settings**.
2. Enable Apple Pay/Google Pay and save changes.
3. Navigate to **WooCommerce > Orders > Add order**.
4. Click on **Add item(s)**, then on **Add product(s)**.
5. Type in some product name, such as "Beanie" and select it.
6. Click on the **Add** button, and finally, on the **Create** button.
7. Copy the **Customer payment page** link.
8. As a customer, navigate to the link you copied.
9. Pay using the ECE and make sure the payment is processed successfully.

**Test: Checkout with ECE in Pay for Order works with WooPay enabled**

1. As a merchant, navigate to **Payments > Settings**.
2. Enable WooPay and save changes.
3. Navigate to **WooCommerce > Orders > Add order**.
4. Click on **Add item(s)**, then on **Add product(s)**.
5. Type in some product name, such as "Beanie" and select it.
6. Click on the **Add** button, and finally, on the **Create** button.
7. Copy the **Customer payment page** link.
8. As a customer, navigate to the link you copied.
9. Pay using the ECE and make sure the payment is processed successfully.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8994` or branch name `add/8869-pay-for-order-support-in-ece` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/8869-pay-for-order-support-in-ece) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f6ab0932f07673de46b7b2678205eb9d5275c7da
- Build time: 2024-07-01 23:07:08 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/express-checkout/event-handlers.js
Comment (lovo-h) [inline client/express-checkout/event-handlers.js]: Can this commented out line be removed? It seems to be unintentionally left behind.

File: client/express-checkout/utils/test/normalize.js
Comment (lovo-h) [inline client/express-checkout/utils/test/normalize.js:304]: Nit: This could also be the following:
```javascript
'wcpay-payment-method': paymentMethodId,
```

File: client/express-checkout/utils/test/normalize.js
Comment (lovo-h) [inline client/express-checkout/utils/test/normalize.js:271]: Nit: Could this be simplified to the following?
```javascript
const event = {
    billingDetails: {},
    shippingAddress: {},
    shippingRate: {},
    expressPaymentType: 'express',
};
```

File: client/express-checkout/event-handlers.js
Comment (lovo-h) [inline client/express-checkout/event-handlers.js]: Nit: This seems to be a duplicate of `onConfirmHandler()`, are we able to remove some of the duplication by using a helper function? Or by passing an `option.<pay_for_order>` that allows for an if-else statement that switches between pay-for-order and create-order.

Although not necessary, it would be nice so that we don't have to maintain two functions.

**Review by lovo-h — CHANGES_REQUESTED** [review]
Code changes are looking good and testing well.

I left a few nit comments, but there were two we may want to handle before merging:
* Unintentionally left behind comment.
* Duplicate code that we may be able to de-duplicate for easier maintenance. This one might be optional.

**Testing Instructions**
1. 🟢 Regression test: Payment request buttons work as expected when ECE is disabled.
2. 🟢 Regression test: Checkout with ECE still works in the standard flows.
3. 🟢 Regression test: Regular checkout works when no express checkout payment method is active.
4. 🟢 Test: Checkout with ECE in Pay for Order works with WooPay disabled.
5. 🟢 Test: Checkout with ECE in Pay for Order works with WooPay enabled.
6. 🟢 `npm run test:js`.

File: client/express-checkout/event-handlers.js
Comment (cesarcosta99) [reply client/express-checkout/event-handlers.js]: Nice catch! Removed in [0005b39](https://github.com/Automattic/woocommerce-payments/pull/8994/commits/0005b3927ee4c50178122ba3b7c7452fc9b703b8).

File: client/express-checkout/utils/test/normalize.js
Comment (cesarcosta99) [reply client/express-checkout/utils/test/normalize.js:271]: It could, but I intentionally left it like that because that's how the event looks like. If we work on passing all that info to the backend so that the order is filled in with it, we'll need to tweak the `expect`, but the event is up-to-date.

File: client/express-checkout/utils/test/normalize.js
Comment (cesarcosta99) [reply client/express-checkout/utils/test/normalize.js:304]: I tend to use literals in one of the sides of the assertion because that's how I avoid being blindsided by the variables. However, since that variable was only being used once, I removed it in [270ff95](https://github.com/Automattic/woocommerce-payments/pull/8994/commits/270ff954e199c5ad1020dc0e95aef03440fad81d)

File: client/express-checkout/event-handlers.js
Comment (cesarcosta99) [reply client/express-checkout/event-handlers.js]: > it would be nice so that we don't have to maintain two functions

Totally agree with that 💯 I refactored this in [fa4559c](https://github.com/Automattic/woocommerce-payments/pull/8994/commits/fa4559c425d6ed640b67b658cfd96ee7e1ec14e8) and the respective tests in [94bef90](https://github.com/Automattic/woocommerce-payments/pull/8994/commits/94bef90c558fae55132f3d32565b8b7a8a712d3d) 👍

**Review by lovo-h — APPROVED** [review]
Code changes look good and test well.

**Testing Instructions**
1. 🟢 Regression test: Payment request buttons work as expected when ECE is disabled.
2. 🟢 Regression test: Checkout with ECE still works in the standard flows.
3. 🟢 Regression test: Regular checkout works when no express checkout payment method is active.
4. 🟢 Test: Checkout with ECE in Pay for Order works with WooPay disabled.
5. 🟢 Test: Checkout with ECE in Pay for Order works with WooPay enabled.
6. 🟢 `npm run test:js` and `npm run test:php`.

Note: For the test `Regression test: Payment request buttons work as expected when ECE is disabled`, I was not able to get the PRBs to load due to the issue reported here:
* p1719971520784039-slack-CU6SYV31A

However, I got things to work as intended by checking out to the commit right before.

**Pull request merged**
